nexus/blueprint-execution/src/resource_allocation.rs

```diff
@@ -0,0 +1,998 @@
+                    && allocated_nic.primary == nic.primary
+                {
+                    info!(
+                        self.log, "found already-allocated NIC";
+                        "zone_type" => zone_type,
+                        "zone_id" => %zone_id,
+                        "nic" => ?allocated_nic,
+                    );
+                    return Ok(true);
+                }
+            }
+
+            warn!(
+                self.log, "zone has unexpected NICs allocated";
+                "zone_type" => zone_type,
+                "zone_id" => %zone_id,
+                "want_nic" => ?nic,
+                "allocated_nics" => ?allocated_nics,
+            );
+
+            bail!(
+                "zone {zone_id} already has {} non-matching NIC(s) allocated",
+                allocated_nics.len()
+            );
+        }
+
+        info!(
+            self.log, "NIC allocation required for zone";
+            "zone_type" => zone_type,
+            "zone_id" => %zone_id,
+            "nid" => ?nic,
+        );
+
+        Ok(false)
+    }
+
+    // Nexus and ExternalDns both use non-SNAT service IPs; this method is used
+    // to allocate external networking for both of them.
+    async fn ensure_external_service_ip(
+        &self,
+        zone_type: &'static str,
+        service_id: Uuid,
+        external_ip: IpAddr,
+        ip_name: &Name,
+    ) -> anyhow::Result<()> {
+        // Only attempt to allocate `external_ip` if it isn't already assigned
+        // to this zone.
+        //
+        // Checking for the existing of the external IP and then creating it
+        // if not found inserts a classic TOCTOU race: what if another Nexus
+        // is running concurrently, we both check and see that the IP is not
+        // allocated, then both attempt to create it? We believe this is
+        // okay: the loser of the race (i.e., the one whose create tries to
+        // commit second) will fail to allocate the IP, which will bubble
+        // out and prevent realization of the current blueprint. That's
+        // exactly what we want if two Nexuses try to realize the same
+        // blueprint at the same time.
+        if self
+            .is_external_ip_already_allocated(
+                zone_type,
+                service_id,
+                external_ip,
+                None,
+            )
+            .await?
+        {
+            return Ok(());
+        }
+        let ip_id = Uuid::new_v4();
+        let description = zone_type;
+        self.datastore
+            .allocate_explicit_service_ip(
+                self.opctx,
+                ip_id,
+                ip_name,
+                description,
+                service_id,
+                external_ip,
+            )
+            .await
+            .with_context(|| {
+                format!(
+                    "failed to allocate IP to {zone_type} {service_id}: \
+                         {external_ip}"
+                )
```

Comment:
Can we add a note here saying that this is expected in case of a race? (Or maybe this is already done at a higher level.)

I'm thinking about future debugging sessions where a simple error in the logs can be confusing.

Reply:
Hmm, I hope the error we get from the datastore would make it clear whether a failure is due to a duplication issue or some other transient db issue. I'm hesitant to label this as "expected if we're racing" when the underlying error might be some other kind of db problem.

Reply:
Okay, just trying to think it through.